include/votca/tools/linalg.h

```diff
@@ -17,238 +17,35 @@
 
 #ifndef __VOTCA_TOOLS_LINALG_H
 #define	__VOTCA_TOOLS_LINALG_H
-#include <votca/tools/votca_config.h>
-#if defined(GSL)
-    #include "votca_gsl_boost_ublas_matrix_prod.h"
-#elif defined(MKL)
-    #include "mkl_boost_ublas_matrix_prod.hpp"
-#endif
-
-#include <boost/numeric/ublas/matrix.hpp>
-#include <boost/numeric/ublas/vector.hpp>
-#include <boost/numeric/ublas/symmetric.hpp>
+#include <votca/tools/eigen.h>
 
 namespace votca { namespace tools {
-    namespace ub = boost::numeric::ublas;
-
-    /**
-     * \brief inverts A
-     * @param A symmetric positive definite matrix
-     * @param V inverse matrix
-     *
-     * This function wraps the inversion of a matrix
-     */
-    void linalg_invert( const ub::matrix<double> &A, ub::matrix<double> &V );
-    void linalg_invert( const ub::matrix<float> &A, ub::matrix<float> &V );
  
     /**
-     * \brief determines Cholesky decomposition of matrix A
-     * @param A symmetric positive definite matrix
-     *
-     * This function wraps the Cholesky decomposition
-     */
-    void linalg_cholesky_decompose( ub::matrix<double> &A );
-    void linalg_cholesky_decompose( ub::matrix<float> &A );
-    /**
-     * \brief solves A*x=b
-     * @param x storage for x
-     * @param A symmetric positive definite matrix for linear system
-     * @param b inhomogeniety
-     * @param if A is not symmetric positive definite throws error code 
-     *
-     * This function wraps the cholesky linear system solver
-     */
-    void linalg_cholesky_solve(ub::vector<double> &x, ub::matrix<double> &A, ub::vector<double> &b);
-
-    /**
-     * \brief solves A*x=b
+     * \brief solves A*x=b under the constraint B*x = 0
      * @param x storage for x
      * @param A matrix for linear equation system
      * @param b inhomogenity
-     * @param residual if non-zero, residual will be stored here
+     * @param constr constrained condition B (or is it the transposed one? check that)
      *
-     * This function wrapps the qrsolver
+     * This function wraps the qrsolver under constraints
      */
-    void linalg_qrsolve(ub::vector<double> &x, ub::matrix<double> &A, ub::vector<double> &b, ub::vector<double> *residual=NULL);
-
-    /**
+    void linalg_constrained_qrsolve(Eigen::VectorXd &x, Eigen::MatrixXd &A,
+            const Eigen::VectorXd &b, const Eigen::MatrixXd &constr);
+  
+  /**
      * \brief solves A*x=b under the constraint B*x = 0
      * @param x storage for x
      * @param A matrix for linear equation system
      * @param b inhomogenity
-     * @param B constrained condition
+     * @param constr constrained condition B (or is it the transposed one? check that)
```

Comment:
This qrsolver is obviously an important function is there anyway we can also provide a link to some documentation that goes into a little more detail about how it works?

Reply:
agreed I did nonly reverse engineer the Eigen implementation and I have no real clue how it works.